theano/tensor/nnet/abstract_conv.py

```diff
@@ -320,13 +339,229 @@ def conv2d_grad_wrt_weights(input,
+    return gradWeight_op(input, output_grad, filter_shape[:-2])
+
+
+def bilinear_kernel_2D(ratio, normalize=True):
+    """Compute 2D kernel for bilinear upsampling
+
+    This function builds the 2D kernel that can be used to upsample
+    a tensor by the given ratio using bilinear interpolation.
+
+    Parameters
+    ----------
+    ratio: int or Constant/Scalar Theano tensor
+        the ratio by which an image will
+        be upsampled by the returned filter in the 2D space.
+
+    normalize: bool
+        param normalize: indicates whether to normalize the
+        kernel or not. Default is True.
+
+    Returns
+    -------
+    symbolic 2D tensor
+        the 2D kernels that can be applied to
+        any given image to upsample it by the indicated ratio
+        using bilinear interpolation in two dimensions.
+
+    """
+
+    hkern = bilinear_kernel_1D(ratio=ratio, normalize=normalize).dimshuffle('x', 0)
+    vkern = bilinear_kernel_1D(ratio=ratio, normalize=normalize).dimshuffle(0, 'x')
+    kern = hkern * vkern
+    return kern
+
+
+def bilinear_kernel_1D(ratio, normalize=True):
+    """Compute 1D kernel for bilinear upsampling
+
+    This function builds the 1D kernel that can be used to upsample
+    a tensor by the given ratio using bilinear interpolation.
+
+    Parameters
+    ----------
+    ratio: int or Constant/Scalar Theano tensor
+        the ratio by which an image is upsampled
+        by the returned filter in the 1D space.
+
+    normalize: bool
+        indicates whether to normalize the
+        kernel or not. Default is True.
+
+    Returns
+    -------
+    symbolic 1D tensor
+        the 1D kernels that can be applied to
+        any given image to upsample it by the indicated ratio
+        using bilinear interpolation in one dimension.
+
+    """
+
+    T = theano.tensor
+    half_kern = T.arange(1, ratio + 1, dtype=theano.config.floatX)
+    kern = T.concatenate([half_kern, half_kern[-2::-1]])
+
+    if normalize:
+        kern /= float(ratio)
+    return kern
+
+
+def bilinear_upsampling(input,
+                        ratio,
+                        batch_size=None,
+                        num_input_channels=None,
+                        use_1D_kernel=True):
+    """Compute bilinear upsampling
+
+    This function will build the symbolic graph for upsampling
+    a tensor by the given ratio using bilinear interpolation.
+
+    Parameters
+    ----------
+    input: symbolic 4D tensor
+        mini-batch of feature map stacks, of shape
+        (batch size, input channels, input rows, input columns)
+        that will be upsampled.
+
+    ratio: integer
+        the ratio by which the input is upsampled
+        in the 2D space (row and col size).
+
+    batch_size: None, int or Constant variable
+        The size of the first dimension of the input variable.
+        Optional, possibly used to choose an optimal implementation.
+        You can give ``None`` for any element of the list to specify that this
```

Comment:
It's not a list. Remove these two lines. Specify instead that `batch_size` will be used only if `num_input_channels` is not None